sql/migrations/warehouse/000032_add_schema_snapshots_to_staging_files.up.sql

```diff
@@ -0,0 +1,27 @@
+-- Create a table to store schema snapshots for staging files.
+-- Each snapshot represents the schema at a specific point in time for a (source, destination, workspace).
+CREATE TABLE IF NOT EXISTS wh_staging_file_schema_snapshots (
+    id UUID PRIMARY KEY DEFAULT gen_random_uuid(), -- Unique identifier for the snapshot
+    schema TEXT NOT NULL,                          -- JSON-encoded schema at the time of the snapshot
+    source_id VARCHAR(64) NOT NULL,                -- Source identifier
+    destination_id VARCHAR(64) NOT NULL,           -- Destination identifier
+    workspace_id VARCHAR(64) NOT NULL,             -- Workspace identifier
+    created_at TIMESTAMP WITHOUT TIME ZONE NOT NULL DEFAULT NOW() -- Snapshot creation timestamp
+);
+
+-- Add columns to wh_staging_files to link each file to a schema snapshot and store a JSON Patch diff.
+ALTER TABLE wh_staging_files
+    ADD COLUMN IF NOT EXISTS schema_snapshot_id UUID REFERENCES wh_staging_file_schema_snapshots(id) ON DELETE SET NULL, -- Link to the schema snapshot
+    ADD COLUMN IF NOT EXISTS schema_patch TEXT; -- JSON Patch (RFC 6902) diff from the snapshot to the actual schema
+
+-- For fast lookup of staging files by schema_snapshot_id.
```

Comment:
For what use-case would we need to look up staging files using snapshot id ? Wouldn't it be the other way around ?

Reply:
This is for archiving, basically, if the snapshot is not being used, we can archive it. Will add comment.